File: frontend_React/src/services/appointmentService.js
import api from './api'

// -- Appointments --

export const getAppointments = async (filters = {}) => {
  const params = new URLSearchParams()
  if (filters.status) params.append('status', filters.status)
  if (filters.priority) params.append('priority', filters.priority)
  if (filters.doctorId) params.append('doctorId', filters.doctorId)
  if (filters.departmentId) params.append('departmentId', filters.departmentId)
  if (filters.fromDate) params.append('fromDate', filters.fromDate)
  if (filters.toDate) params.append('toDate', filters.toDate)
  if (filters.patientId) params.append('patientId', filters.patientId)
  
  const response = await api.get(`/api/appointments?${params.toString()}`)
  return response.data
}

export const getAppointmentById = async (id) => {
  const response = await api.get(`/api/appointments/${id}`)
  return response.data
}

export const bookAppointment = async (data) => {
  const response = await api.post('/api/appointments', data)
  return response.data
}

export const updateAppointment = async (id, data) => {
  const response = await api.put(`/api/appointments/${id}`, data)
  return response.data
}

export const cancelAppointment = async (id, reason) => {
  const response = await api.post(`/api/appointments/${id}/cancel`, { reason })
  return response.data
}

export const rescheduleAppointment = async (id, data) => {
  const response = await api.post(`/api/appointments/${id}/reschedule`, data)
  return response.data
}

export const getAppointmentHistory = async (id) => {
  const response = await api.get(`/api/appointments/${id}/history`)
  return response.data
}

export const confirmEmergency = async (id) => {
  const response = await api.post(`/api/appointments/${id}/confirm-emergency`)
  return response.data
}

// -- Queues --

export const getQueue = async (doctorId) => {
  const response = await api.get(`/api/queues/${doctorId}`)
  return response.data
}

export const callQueueEntry = async (id) => {
  const response = await api.post(`/api/queues/${id}/call`)
  return response.data
}

export const checkIn = async (appointmentId) => {
  const response = await api.post(`/api/queues/check-in/${appointmentId}`)
  return response.data
}

export const markNoShow = async (id) => {
  const response = await api.post(`/api/queues/${id}/no-show`)
  return response.data
}

export const markCompleted = async (id) => {
  const response = await api.post(`/api/queues/${id}/complete`)
  return response.data
}

export const getQueueEntryStatus = async (id) => {
  const response = await api.get(`/api/queues/${id}/status`)
  return response.data
}

// -- Availability --

export const getAvailableSlots = async (date, doctorId, departmentId) => {
  const params = new URLSearchParams()
  params.append('date', date)
  if (doctorId) params.append('doctorId', doctorId)
  if (departmentId) params.append('departmentId', departmentId)
  
  const response = await api.get(`/api/availability/slots?${params.toString()}`)
  return response.data
}

export const getRescheduleSuggestions = async (doctorId, preferredDate, durationMinutes = 30) => {
  const params = new URLSearchParams()
  params.append('doctorId', doctorId)
  params.append('preferredDate', preferredDate)
  params.append('durationMinutes', durationMinutes)
  
  const response = await api.get(`/api/availability/suggestions?${params.toString()}`)
  return response.data
}

// -- Notifications --

export const getNotifications = async () => {
  const response = await api.get('/api/notifications')
  return response.data
}

export const markNotificationRead = async (id) => {
  const response = await api.post(`/api/notifications/${id}/read`)
  return response.data
}

File: frontend_Flutter/lib/services/appointment_service.dart
import '../../core/network/api_client.dart';
import '../models/appointments/appointment_model.dart';
import '../models/appointments/appointment_slot_model.dart';
import '../models/appointments/appointment_history_model.dart';
import '../models/appointments/queue_entry_model.dart';
import '../models/appointments/doctor_summary_model.dart';

class AppointmentService {
  final ApiClient _apiClient;

  AppointmentService(this._apiClient);

  // --- Appointment endpoints ---

  Future<List<AppointmentModel>> getMyAppointments({
    String? status,
    String? fromDate,
    String? toDate,
  }) async {
    final queryParts = <String>[];
    if (status != null) queryParts.add('status=$status');
    if (fromDate != null) queryParts.add('fromDate=$fromDate');
    if (toDate != null) queryParts.add('toDate=$toDate');
    final query = queryParts.isEmpty ? '' : '?${queryParts.join('&')}';
    final response = await _apiClient.get('/appointments$query');
    if (response is List) {
      return response
          .map((e) => AppointmentModel.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  Future<AppointmentModel> getAppointmentById(int id) async {
    final response = await _apiClient.get('/appointments/$id');
    return AppointmentModel.fromJson(response as Map<String, dynamic>);
  }

  Future<List<AppointmentHistoryModel>> getAppointmentHistory(int id) async {
    try {
      final response = await _apiClient.get('/appointments/$id/history');
      if (response is List) {
        return response
            .map((e) =>
                AppointmentHistoryModel.fromJson(e as Map<String, dynamic>))
            .toList();
      }
    } catch (_) {
      // History may not be available — fail silently
    }
    return [];
  }

  Future<AppointmentModel> bookAppointment({
    required int doctorId,
    required int appointmentType,
    required String scheduledStart,
    required int estimatedDurationMinutes,
    required int priority,
    String? notes,
  }) async {
    final response = await _apiClient.post('/appointments', data: {
      'doctorId': doctorId,
      'appointmentType': appointmentType,
      'scheduledStart': scheduledStart,
      'estimatedDurationMinutes': estimatedDurationMinutes,
      'priority': priority,
      if (notes != null && notes.isNotEmpty) 'notes': notes,
    });
    return AppointmentModel.fromJson(response as Map<String, dynamic>);
  }

  Future<void> cancelAppointment(int id, String reason) async {
    await _apiClient.post('/appointments/$id/cancel',
        data: {'reason': reason});
  }

  Future<AppointmentModel> rescheduleAppointment(
    int id, {
    required String newScheduledStart,
    required int newEstimatedDurationMinutes,
    required String reason,
  }) async {
    final response = await _apiClient.post('/appointments/$id/reschedule', data: {
      'newScheduledStart': newScheduledStart,
      'newEstimatedDurationMinutes': newEstimatedDurationMinutes,
      'reason': reason,
    });
    return AppointmentModel.fromJson(response as Map<String, dynamic>);
  }

  // --- Availability endpoint ---

  Future<List<AppointmentSlotModel>> getAvailableSlots({
    int? doctorId,
    String? date,
  }) async {
    final queryParts = <String>[];
    if (doctorId != null) queryParts.add('doctorId=$doctorId');
    if (date != null) queryParts.add('date=$date');
    final query = queryParts.isEmpty ? '' : '?${queryParts.join('&')}';
    final response = await _apiClient.get('/availability/slots$query');
    if (response is List) {
      return response
          .map((e) =>
              AppointmentSlotModel.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  // --- Queue endpoints ---

  Future<List<QueueEntryModel>> getQueueForDoctor(int doctorId) async {
    final response = await _apiClient.get('/queues/$doctorId');
    if (response is List) {
      return response
          .map((e) => QueueEntryModel.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }

  // --- Doctor search (reads from Users endpoint, Doctor role) ---
  // TODO: Replace with shared Doctor module endpoint once available.

  Future<List<DoctorSummaryModel>> searchDoctors({String? query}) async {
    final q = query != null && query.isNotEmpty ? '?search=$query&role=Doctor' : '?role=Doctor';
    final response = await _apiClient.get('/Users$q');
    if (response is List) {
      return response
          .map((e) => DoctorSummaryModel.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    return [];
  }
}
